PowerEditor/src/WinControls/ProjectPanel/Directory.cpp

```diff
@@ -0,0 +1,450 @@
+// following:                                                             
+// 1. Integrates source code from Notepad++.
+// 2. Integrates/includes/aggregates Notepad++ into a proprietary executable
+//    installer, such as those produced by InstallShield.
+// 3. Links to a library or executes a program that does any of the above.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+
+#include "Directory.h"
+#include <Shlwapi.h>
+
+Directory::Directory(bool hideEmptyDirs)
+	: _exists(false)
+	, _hideEmptyDirs(hideEmptyDirs)
+	, _wasRead(false)
+{
+	_lastWriteTime.dwLowDateTime = 0;
+	_lastWriteTime.dwHighDateTime = 0;
+	enablePrivileges();
+	_filters.push_back(generic_string(TEXT("*.*")));
+}
+
+Directory::Directory(const generic_string& path, const std::vector<generic_string>& filters, bool hideEmptyDirs, bool autoread)
+	: _exists(false)
+	, _path(path)
+	, _filters(filters)
+	, _hideEmptyDirs(hideEmptyDirs)
+	, _wasRead(false)
+{
+	_lastWriteTime.dwLowDateTime = 0;
+	_lastWriteTime.dwHighDateTime = 0;
+	enablePrivileges();
+	if (autoread)
+		read(path);
+}
+
+bool Directory::read(const generic_string& path, const std::vector<generic_string>& filters)
+{
+	_path = path;
+	_exists = false;
+	_lastWriteTime.dwLowDateTime = 0;
+	_lastWriteTime.dwHighDateTime = 0;
+	_files.clear();
+	_dirs.clear();
+	_invisibleDirs.clear();
+
+	if (&_filters != &filters)
+		_filters = filters;
+
+	// first, read directories, never filtered.
+	append(_path, TEXT("*.*"), true);
+
+	// then read files, by each filter
+	if (_filters.empty())
+		append(_path, TEXT("*.*"), false);
+	else
+		for (size_t i=0; i<_filters.size(); ++i)
+			append(_path, _filters[i], false);
+
+
+	readLastWriteTime(_lastWriteTime);
+
+	_wasRead = true;
+
+	return _exists;
+
+}
+
+bool Directory::readIfChanged(bool respectEmptyDirs)
+{
+	if (!_wasRead)
+	{
+		read();
+		return true;
+	}
+
+	if (respectEmptyDirs && _hideEmptyDirs)
+	{
+		if (!writeTimeHasChanged() && !containsDataChanged())
+			return false;
+	}
+	else
+	{
+		if (!writeTimeHasChanged())
+			return false;
+	}
+
+	Directory cmp(_path,_filters);
+	if (*this != cmp)
+	{
+		*this = cmp;
+		return true;
+	}
+	return false;
+}
+
+void Directory::append(const generic_string& path, const generic_string& filter, bool readDirs)
+{
+	bool readFiles = !readDirs;
+
+	generic_string searchPath = _path + TEXT("\\") + filter;
+
+	WIN32_FIND_DATAW fd;
+
+	HANDLE hFind = FindFirstFileEx(searchPath.c_str(), FindExInfoBasic, &fd, FindExSearchNameMatch, NULL, FIND_FIRST_EX_LARGE_FETCH);
```

Comment:
*imagine that this comment is on the line*:

```
if (fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
```
...*because my phone is misbehaving*.

Make this explicit? i.e.:

```
if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
```